### ISSUE TITLE
Model doesnot converge even after trying everything

### ISSUE THREAD

**Issue by Luv-crypto** [lululxvi/deepxde#1739]
### Discussed in https://github.com/lululxvi/deepxde/discussions/1738

<div type='discussions-op-text'>

<sup>Originally posted by **Luv-crypto** May 14, 2024</sup>
Hello, 
I am trying to solve this pde  
![Screenshot 2024-05-14 132943](https://github.com/lululxvi/deepxde/assets/75800998/65e43c35-ad60-4a30-88cb-3964692fbc34)

with Drichlet boundary conditions for a rectangular domain from -1 to +1. but even after trying different combinations of network architectures i.e deeper and wider networks, different NNs such as FNN, ResNets and different combinations of optimizers and different distributions of points I cannot get the model to predict better. Can someone help me below is the code

import deepxde as dde
import numpy as np
import tensorflow as tf  

def pde(x, y):
    """ Function to compute PDE residual """
    dy_xx = dde.grad.hessian(y, x, component=0, i=0, j=0)
    dy_yy = dde.grad.hessian(y, x, component=0, i=1, j=1)
    # The source term S, using TensorFlow operations
    S = -(2 + tf.square(np.pi) * x[:, 0:1] * (1 - x[:, 0:1])) * tf.cos(tf.constant(np.pi) * x[:, 1:2])
    return dy_xx + dy_yy - S  

# Define the computational domain
geom = dde.geometry.Rectangle([-1, -1], [1, 1])

# Define hard enforcement for the boundary conditions using output transformation

def output_transform(x, output):
    """ Custom output transformation function for enforcing hard boundary conditions. """
    # Extract x and y coordinates
    x_coord = x[:, 0:1]  # x coordinate
    y_coord = x[:, 1:2]  # y coordinate
    
    # Boundary conditions
    phi_x1 = 0 * tf.ones_like(output)  # phi = 0 at x = 1
    phi_x_1 = -2 * tf.cos(np.pi * y_coord)  # phi = -2cos(pi*y) at x = -1
    phi_y1_y_1 = -x_coord * (1 - x_coord)  # phi = -x(1-x) at y = 1 and y = -1
    
    # Apply boundary conditions
    # Using tf.where to selectively apply boundary conditions
    output = tf.where(tf.equal(x_coord, 1), phi_x1, output)
    output = tf.where(tf.equal(x_coord, -1), phi_x_1, output)
    output = tf.where(tf.logical_or(tf.equal(y_coord, 1), tf.equal(y_coord, -1)), phi_y1_y_1, output)
    
    return output

def func(x):
    return x[:,0:1]*(1-x[:,0:1])*np.cos(np.pi*x[:,1:])

data = dde.data.PDE(geom, pde, [], num_domain=20000, num_boundary=1000 , solution= func , num_test=2000)  # No explicit boundary conditions are passed here

!# net = dde.maps.ResNet(input_size=2, output_size=1, num_neurons=50, num_blocks=num_res_blocks, activation=activation, kernel_initializer=kernel_initializer)
net = dde.maps.FNN([2] + [100]*10 + [1], "tanh", "Glorot normal")
# Apply the output transformation to the neural network
net.apply_output_transform(output_transform)

# Combine the PDE and the boundary conditions

!# resampler = dde.callbacks.PDEPointResampler(period=10)
# Model
model = dde.Model(data, net)

# Train the model
model.compile("adam", lr=1e-3, metrics=["l2 relative error"])
model.train(iterations=8000)

model.compile("L-BFGS-B", metrics=["l2 relative error"])
losshistory, train_state = model.train()

# Plotting the results
dde.saveplot(losshistory, train_state, issave=True, isplot=True)

![100_10_FNN](https://github.com/lululxvi/deepxde/assets/75800998/b5d814f9-8305-4fdc-847c-5b0723bd0fd9)
Here the what the solution should look like. The exact solution is x*(1-x)*cos(pi*y)</div>

**Comment by ShahedMohammed** [comment]
It seems that this PINN model applies the hard constraint only on the boundary without changing the domain data. Maybe something like this would work for the hard constraint $$\((\phi x)(x+1)-2cos(\pi y))(y+1)(y-1)-x(1-x)$$.

**Comment by Luv-crypto** [comment]
Hi @ShahedMohammed  thankyou for the quick response can you please explain how does this function will cover my boundary conditions. I dont get it 
my conditions are
    For x=1x=1 and all y the function should yield 0.
    For x=−1 and all y, the function should yield −2cos⁡(πy)
    For y=1 or y=−1 and all x, the function should yield −x(1−x).
    
    Furthermore, the training loss doesnot decrease but the test loss decreases
    'build' took 5.286763 s

'compile' took 33.948308 s

Training model...

Step      Train loss    Test loss     Test metric   
0         [2.17e+01]    [1.97e+01]    [1.06e+00]    
1000      [1.31e+00]    [5.16e-02]    [2.97e+00]    
2000      [1.33e+00]    [6.06e-02]    [2.80e+00]    
3000      [1.28e+00]    [1.28e-02]    [2.69e+00]    
4000      [1.30e+00]    [3.75e-02]    [2.60e+00]    
5000      [1.27e+00]    [1.54e-03]    [2.53e+00]    
6000      [1.26e+00]    [1.03e-03]    [2.25e+00]    
7000      [1.28e+00]    [1.41e-02]    [2.22e+00]    
8000      [1.27e+00]    [2.29e-03]    [1.95e+00]    

Best model at step 6000:
  train loss: 1.26e+00
  test loss: 1.03e-03
  test metric: [2.25e+00]

'train' took 1938.201845 s

Compiling model...
'compile' took 2.830205 s

Training model...

Step      Train loss    Test loss     Test metric   
8000      [1.27e+00]    [2.29e-03]    [1.95e+00]    
INFO:tensorflow:Optimization terminated with:
  Message: CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
  Objective function value: 1.263703
  Number of iterations: 162
  Number of functions evaluations: 184
8184      [1.26e+00]    [4.60e-05]    [7.98e-01]    

Best model at step 8184:
  train loss: 1.26e+00
  test loss: 4.60e-05
  test metric: [7.98e-01]

'train' took 270.986128 s

![image](https://github.com/lululxvi/deepxde/assets/75800998/b0721879-7adf-47a3-ad3e-3985a5006e0b)
![image](https://github.com/lululxvi/deepxde/assets/75800998/da150bb2-7c8f-4817-8a5c-b4380a0ef684)

**Comment by ShahedMohammed** [comment]
Hi @Luv-crypto, My solution was wrong. It did not satisfy all the boundary conditions and I am deleting it.

**Comment by Luv-crypto** [comment]
Thankyou for the suggestion. I will try to do that. Do you have any idea why the training loss doesnot decrease eventhough i tried with deeper and wider networks with different optimizers?

**Comment by Luv-crypto** [comment]
Hi @lululxvi sorry to disturb you. But could me take a look whenever you have time
